ramips: add support for Senao Engenius EPG600

mcprat opened the pull request:
original issues:

- ~~depends on #4111~~
- 5 GHz wireless not working yet #8780
- ~~in linux 5.10, the flash chip will have 4k blocksize [1], for now uboot-env is read-only~~

Not supported:

- there are 2 phone ports for VoIP with Si3050 + Si3019 chipset [2]

[1] https://github.com/torvalds/linux/commit/02892d405358288e0057cbdfe6476d13ea62fd2f
[2] https://www.silabs.com/documents/public/data-sheets/Si3050-11-18-19.pdf

mcprat committed:
ramips: add support for Senao Engenius EPG600

FCC ID: A8J-EPG600

Engenius EPG600 is an indoor wireless router with
1 Gb ethernet switch, dual-band wireless,
internal antenna plates, USB, and phone lines (not supported)

this board is a Senao device:
the hardware is equivalent to EnGenius ESR600 (except for phone lines)
the software is Senao SDK which is based on openwrt and uboot
which uses the legacy Senao header with Vendor / Product IDs
to verify the firmware upgrade image.

**Specification:**

  - MT7620 SOC		MIPS 24kec, 2.4 GHz WMAC, 2x2
  - RT5592N WLAN	PCI chip, 5 GHz, 2x2
  - QCA8337N Gb SW	RGMII GbE, SW P0 -- SOC P5, 5 LEDs
  - 40 MHz clock
  - 16 MB FLASH		MX25L12845EMI-10G
  - 64 MB RAM		NT5TU32M16
  - UART console	J2, populated
  - USB 2.0 port	direct to SOC
  - 6 GPIO LEDs		power, 2G, 5G, wps2g, wps5g, line
  - 3 buttons		reset, wps, "reg" (registeration)
  - 4 antennas		internal omni-directional plates

NOT YET SUPPORTED: VoIP

  - Si3050-FT + Si3019-FT	Voice DAA, SPI control, PCM data
  - Phone Ports "TEL", "LINE"	RJ11, 4P2C (2 pins)

**MAC addresses:**

  MAC address labeled as MAC ADDRESS
  MACs present in both wifi cal data and uboot environment

  eth0.1/phy1	----	*:82	rf 0x4
  phy0		----	*:83	factory 0x4
  eth0.2	MAC	*:b8	"wanaddr"

**Installation:**

  Method 1: Firmware upgrade page:

    (if you cannot access the APs webpage)
    factory reset with the reset button
    connect ethernet to a computer
    OEM webpage at 192.168.0.1
    username and password 'admin'

    Navigate to gear icon, "Device Management", "Tools"
    select the factory.dlf image
    Upload and verify checksum

  Method 2: Serial to upload initramfs:

    Follow directions for TFTP recovery
    upload and boot initramfs and do a sysupgrade

**TFTP recovery:**

  Requires UART serial console, reset button does nothing

  rename initramfs-kernel.bin to 'uImageEPG600'
  make available on TFTP server at 192.168.99.8
  power board, interrupt boot with "4"
  execute `tftpboot` and `bootm` (with the load address)

**Return to OEM:**

  Images from OEM are provided, but not compatible
  with openwrt sysupgrade. So it must be modified.

  Alternatively, back up all mtd partitions before flashing

**Note on switch registers:**

  The necessary registers needed for the QCA8337 switch
  can be read from interrupted boot (tftpboot, bootm)
  by using the following lines in the switch driver ar8327.c
  in the function 'ar8327_hw_config_of'
  where 'qca,ar8327-initvals' is parsed from DTS
  before the new register values are written:

    pr_info("0x04 %08x\n", ar8xxx_read(priv, AR8327_REG_PAD0_MODE));
    pr_info("0x08 %08x\n", ar8xxx_read(priv, AR8327_REG_PAD5_MODE));
    pr_info("0x0c %08x\n", ar8xxx_read(priv, AR8327_REG_PAD6_MODE));
    pr_info("0x10 %08x\n", ar8xxx_read(priv, AR8327_REG_POWER_ON_STRAP));

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
@polex use my DTS here to make your DTS for master

polex commented:
On Sitecom WLR-4100, just one RGMII bus connects MT7620 (port 5) with AR8x37 (port 0).
pinctrl-0 = <&mdio_pins &rgmii1_pins>;
It may be the same for EPG600.

polex commented:
@mpratt14 
Does USB 2.0 works in high speed mode?
WLR 4100 fails in high speed mode.
The failure triggers a power cycle which disables USB tethering.
That is why i removed ehci section. At least everythings works at USB full speed.

mpratt14 commented:
> On Sitecom WLR-4100, just one RGMII bus connects MT7620 (port 5) with AR8x37 (port 0).
> pinctrl-0 = <&mdio_pins &rgmii1_pins>;

for me, both the 2 RGMII interfaces show link up
and even if this is not the case, it doesn't hurt to set the pin function anyway
(unless the GPIO is used for something else)


> Does USB 2.0 works in high speed mode?

I have connected a USB 2 device and as far as I can tell it works

```
...
...
...
[    7.911005] ehci-platform 101c0000.ehci: EHCI Host Controller
[    7.916921] ehci-platform 101c0000.ehci: new USB bus registered, assigned bus number 1
[    7.925146] ehci-platform 101c0000.ehci: irq 26, io mem 0x101c0000
[    7.936110] ehci-platform 101c0000.ehci: USB 2.0 started, EHCI 1.00
[    7.943572] hub 1-0:1.0: USB hub found
[    7.947941] hub 1-0:1.0: 1 port detected
[    7.957932] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    7.965660] ohci-platform: OHCI generic platform driver
[    7.971385] ohci-platform 101c1000.ohci: Generic Platform OHCI controller
[    7.978399] ohci-platform 101c1000.ohci: new USB bus registered, assigned bus number 2
[    7.986608] ohci-platform 101c1000.ohci: irq 26, io mem 0x101c1000
[    8.018141] hub 2-0:1.0: USB hub found
[    8.022514] hub 2-0:1.0: 1 port detected
[    8.036203] usbcore: registered new interface driver usb-storage
...
...
...
[   99.392353] usb 1-1: new high-speed USB device number 2 using ehci-platform
[  106.234116] usb 1-1: USB disconnect, device number 2
```

adschm commented on target/linux/ramips/mt7620/base-files/etc/board.d/02_network:
```diff
@@ -300,9 +301,11 @@ ramips_setup_macs()
 	engenius,esr600)
 		lan_mac=$(mtd_get_mac_ascii u-boot-env ethaddr)
 		wan_mac=$(mtd_get_mac_ascii u-boot-env wanaddr)
+		label_mac=$wan_mac
```
Changing this for esr600 should be a separate patch.

polex commented:
> I have connected a USB 2 device and as far as I can tell it works

I have identical USB  messages in boot, but I get:
```
...
...
[ 1066.007540] usb 1-1: new high-speed USB device number 2 using ehci-platform
[ 1066.142423] usb 1-1: device descriptor read/64, error -71
[ 1066.317379] usb 1-1: device descriptor read/64, error -71
[ 1066.490294] usb 1-1: new high-speed USB device number 3 using ehci-platform
[ 1066.624884] usb 1-1: device descriptor read/64, error -71
[ 1066.801440] usb 1-1: device descriptor read/64, error -71
[ 1066.846529] usb usb1-port1: attempt power cycle
[ 1067.094513] usb 1-1: new high-speed USB device number 4 using ehci-platform
[ 1067.254412] usb 1-1: device not accepting address 4, error -71
[ 1067.386061] usb 1-1: new high-speed USB device number 5 using ehci-platform
[ 1067.546321] usb 1-1: device not accepting address 5, error -71
[ 1067.552354] usb usb1-port1: unable to enumerate USB device
[ 1067.697322] usb 2-1: new full-speed USB device number 2 using ohci-platform
[ 1067.794652] usb 2-1: not running at top speed; connect to a high speed hub
...
...
```
after connection.

May it be this the problem?
```
	gpio_export {
		compatible = "gpio-export";
		#size-cells = <0>;
		usb-power {
			gpio-export,name = "usb-power";
			gpio-export,output = <1>;
			gpios = <&gpio1 14 GPIO_ACTIVE_HIGH>;
		};
	};

```

mpratt14 commented:
it is possible `gpio_export` node is not needed...

you should test with a usb device that shows if it has power with an LED
it may also be possible that it is the wrong GPIO number

mcprat committed:
ramips: add label MAC address for EPG600 and ESR600

set label mac for the current ESR600 and new EPG600
using the board.d scripts

Signed-off-by: Michael Pratt <mcpratt@pm.me>

polex commented:
> you should test with a usb device that shows if it has power with an LED

With my big surprise, another USB key with LED worked at high speed. 
```
[36384.001036] usb 1-1: new high-speed USB device number 7 using ehci-platform
[36384.118130] usb-storage 1-1:1.0: USB Mass Storage device detected
[36384.135556] scsi host0: usb-storage 1-1:1.0
[36384.728151] scsi 0:0:0:0: Direct-Access     Generic  Flash Disk       8.07 PQ: 0 ANSI: 2
[36384.741344] sd 0:0:0:0: [sda] 7864320 512-byte logical blocks: (4.03 GB/3.75 GiB)
[36384.749769] sd 0:0:0:0: [sda] Write Protect is off
[36384.754675] sd 0:0:0:0: [sda] Mode Sense: 03 00 00 00
[36384.760586] sd 0:0:0:0: [sda] No Caching mode page found
[36384.766106] sd 0:0:0:0: [sda] Assuming drive cache: write through
[36384.777169]  sda: sda1
[36384.784521] sd 0:0:0:0: [sda] Attached SCSI removable disk
[36389.013339] usb 1-1: USB disconnect, device number 7
```

So I tried an USB 3.1 SSD

```
[36475.305733] usb 1-1: new high-speed USB device number 8 using ehci-platform
[36475.546193] usb 1-1: device descriptor read/64, error -71
[36476.324629] usb 1-1: device descriptor read/64, error -71
[36477.110647] usb 1-1: new high-speed USB device number 9 using ehci-platform
[36477.785276] usb 1-1: device descriptor read/64, error -71
[36478.330340] usb 1-1: device descriptor read/64, error -71
[36478.401133] usb usb1-port1: attempt power cycle
[36479.548232] usb usb1-port1: Cannot enable. Maybe the USB cable is bad?
[36480.057805] usb 1-1: new high-speed USB device number 11 using ehci-platform
[36480.278536] usb 1-1: device not accepting address 11, error -71
[36480.284695] usb usb1-port1: unable to enumerate USB device
[36480.540989] usb 2-1: new full-speed USB device number 3 using ohci-platform
[36480.683722] usb 2-1: not running at top speed; connect to a high speed hub
[36480.706815] usb-storage 2-1:1.0: USB Mass Storage device detected
[36480.724809] scsi host0: usb-storage 2-1:1.0
[36481.373005] scsi 0:0:0:0: Direct-Access     WD       My Passport 25F3 1012 PQ: 0 ANSI: 6
[36481.387110] scsi 0:0:0:1: Enclosure         WD       SES Device       1012 PQ: 0 ANSI: 6
[36481.415244] sd 0:0:0:0: [sda] 500051968 512-byte logical blocks: (256 GB/238 GiB)
[36481.425701] sd 0:0:0:0: [sda] Write Protect is off
[36481.430763] sd 0:0:0:0: [sda] Mode Sense: 47 00 10 08
[36481.439812] sd 0:0:0:0: [sda] No Caching mode page found
[36481.445353] sd 0:0:0:0: [sda] Assuming drive cache: write through
[36481.511004]  sda: sda1 sda2 < sda5 > sda3
[36481.529014] sd 0:0:0:0: [sda] Attached SCSI disk
[36488.331082] usb 2-1: USB disconnect, device number 3
```
Connected with factory cable, Finally it may be an weak power supply issue.

mpratt14 commented:
Yes, it is likely that the SSD requires too much power to run, and this is causing the voltage to drop on the data line or on the device board, which results in the errors you see (not able to communicate with the USB interface of the device)

polex commented:
I listed sitecom,wlr-4100 just in ramips_setup_interfaces() function in network_02. I did not add in ramips_setup_macs() function and I did not in package/boot/uboot-envtools/files/ramips file.
Ifconfig shows correctly both MAC factory addreess.
What is the purpose of this labelling?

mpratt14 commented:
it depends on the device

> ramips_setup_macs()

this is used when it is not possible to set MAC address using DTS

> package/boot/uboot-envtools/files/ramips

this is used when flash has a partition for uboot environment variables and it is necessary or useful to access it

mpratt14 commented:
bug report for 5 GHz not working

https://bugs.openwrt.org/index.php?do=details&task_id=3765

adschm commented:
While I'm aware this is a draft, note that the final version will need specs and flashing instructions in the commit message.

mcprat committed:
ramips: mt7620: split gsw hw_init to mac_init, ephy_init

in order for the option ephy-disable to work
without also needing ephy-base option,
we have to skip all the lines that write to mdio addresses that
assume those addresses do not have an external switch.

Otherwise, ephy ports will be disabled in hardware,
but register writes still happen as if they are enabled.

Split the functions so that other things are done first,
and ephy port setup can be skipped with a simple "return".

Tested on Engenius EPG600 (MT7620A ver:2 eco:3)
with QCA8337 external switch

Ref: cc6fd6fbb505 ("ramips: mt7620: add ephy-disable option to switch driver")
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
@adschm @hauke this is no longer WIP, ready for review

@981213 ping, if you have time to review small code change here, thanks in advance.

mpratt14 commented:
small syntax error fixed

reordered commits

mpratt14 commented:
realized that IRQ_NONE causes a kernel trace, so handled outside of the function instead

mpratt14 commented:
nevermind, not going to touch the irq in this PR

dangowrt commented:
This needs to be converted to use `nvmem-cells` in device tree for MAC addresses from MTD.

mpratt14 commented:
> This needs to be converted to use `nvmem-cells` in device tree for MAC addresses from MTD.

@dangowrt I was about to rebase and do this and push but I realized it's done already, unless I am missing something...

mpratt14 commented:
rebased and run tested, status is the same (everything works except 5 GHz wireless)

but theres a strange message in kernel log

```
[    0.839298] spi spi0.0: force spi mode3
[    0.844688] spi-nor spi0.0: mx25l12805d (16384 Kbytes)
[    0.850114] 7 fixed-partitions partitions found on MTD device spi0.0
[    0.856678] OF: Bad cell count for /palmbus@10000000/spi@b00/flash@0/partitions
[    0.864162] OF: Bad cell count for /palmbus@10000000/spi@b00/flash@0/partitions
[    0.872287] OF: Bad cell count for /palmbus@10000000/spi@b00/flash@0/partitions
[    0.879831] OF: Bad cell count for /palmbus@10000000/spi@b00/flash@0/partitions
[    0.887556] Creating 7 MTD partitions on "spi0.0":
```

@hauke is this mergable or do you want me to split the change to the driver to a separate PR?

mpratt14 commented:
@tmn505 would you have a moment to help me figure out the shell password for this board?

I read your commit for EPG5000 and that shell password works for the similar boards, ESR900, ESR1200, ESR1750, but this one is different 

did you decompile their busybox or did you find a script with it?

tmn505 commented:
I took the password from OpenWrt wiki, so no reverse engineering from my side. Judging by wiki history it was inputted by bk138. But that password seem to have longer history. It first appeared in ESR900 wiki page inputted by cyrozap. There seem to be another password which allows to enter shell. Both passwords come from GPL archive from that site. First is defined in `GPL_RELEASE/ISD2/configs/h/gconfig.h` as configuration option, second is in `GPL_RELEASE/ISD2/TARGET_RAMDISK/MIPS32_STAGE/sbin/config_term` binary as a string. If You have flash dump of the device maybe it's stored in the same binary. Have You tried some obvious ones like 'password' or '1234'? The other device I ported from Senao was a Sitecom router, there it was 'SitecomSenao', maybe here it's 'EngeniusSenao' or other capitalization, like 'enGeniusSenao'.

mpratt14 commented:
thanks for the tip, I looked at ESR900 wiki page and tried the alternate `lin17` and that worked

mpratt14 commented:
i was hoping to get access to a real kernel log through the shell, but it's so primitive... I don't think it was built with the kernel options for printk...

```
# dmesg
dmesg: klogctl: Function not implemented
```

tmn505 commented:
I think the same was with EPG5000, fortunately all I needed was to port it from ar71xx, as bk138 has done most of the work there. Maybe devmem is enabled or debugfs, which could help, the last option would be asking for source code (but I think You've already done that).

hauke commented:
> rebased and run tested, status is the same (everything works except 5 GHz wireless)
> 
> but theres a strange message in kernel log
> 
> ```
> [    0.839298] spi spi0.0: force spi mode3
> [    0.844688] spi-nor spi0.0: mx25l12805d (16384 Kbytes)
> [    0.850114] 7 fixed-partitions partitions found on MTD device spi0.0
> [    0.856678] OF: Bad cell count for /palmbus@10000000/spi@b00/flash@0/partitions
> [    0.864162] OF: Bad cell count for /palmbus@10000000/spi@b00/flash@0/partitions
> [    0.872287] OF: Bad cell count for /palmbus@10000000/spi@b00/flash@0/partitions
> [    0.879831] OF: Bad cell count for /palmbus@10000000/spi@b00/flash@0/partitions
> [    0.887556] Creating 7 MTD partitions on "spi0.0":
> ```
> 
> @hauke is this mergable or do you want me to split the change to the driver to a separate PR?

No it is not a problem.

hauke commented:
Thank you for the patches, I applied them to master in 6b07d0c077e8bf98ca7e49864774b864e581205a, 4ef86c620f03b3b8b3220d11f77948e7a10469d8 and 6f6d24a00ad43d82a52580d2610a1fd90cedd60a.
